splunk-adx-alert-addon/adx_connector.py

```diff
@@ -0,0 +1,205 @@
+                        stream = io.BytesIO(json_bytes)
+                        stream_descriptor = StreamDescriptor(stream)
+                        self.client.ingest_from_stream(stream_descriptor, ingestion_properties=self.ingestion_props)
+                        self.successfully_sent_events_number += 1
+                        break # Event posted to ADX, break out of the retry loop
+                    except Exception as ex:
+                        helper.log_error('Session= {} Error while sending data to ADX. {}'.format(self.session_identifier,ex))
+                        if isinstance(ex, KustoApiError) and ex.get_api_error().permanent:
+                            helper.log_error('Session= {} Error while sending data to ADX due to KustoApiError - Permanent Error. {}'.format(self.session_identifier,ex))
+                            break; # Permanent error, break out of the retry loop
+                        if isinstance(ex, KustoMultiApiError) and ex.get_api_errors()[0].permanent:
+                            helper.log_error('Session= {} Error while sending data to ADX due to KustoMultiApiError - Permanent Error {}'.format(self.session_identifier,ex))
+                            break; # Permanent error, break out of the retry loop
+                        retry_count += 1
+                        helper.log_error('Session= {} Error while sending data to ADX. Retrying... {}'.format(self.session_identifier,ex))
+                        time.sleep(retry_interval * (2 ** retry_count) + random.randint(1,10)) # Exponential backoff with jitter
+                else:
+                    helper.log_error('Session= {} . Failed to ingest event even after {} retries'.format(self.session_identifier,self.retry_times))   
+        self.ingest_to_adx(helper)
+
+    def ingest_to_adx(self, helper):
+        #if durable mode is disabled, then return
+        if(helper.get_param('durable_mode') == "0"):
+            return 0
+        time.sleep(2) #wait for 2 sec before ingesting to ADX
+        # Set the folder path to read the JSON files 
+        folder_path = os.getcwd()
+        buffer_folder_name = os.path.join(folder_path, "ADX_Ingest")
+
+        # Iterate through the JSON files in the directory
+        for file in self._get_files_not_modified(buffer_folder_name):
+            file_path = os.path.join(buffer_folder_name, file)
+            helper.log_info("Processing file = {}".format(file_path))
+            if file.endswith('.json.gz') and self._is_file_not_being_modified(file_path) :
+                # ingest from file
+                file_descriptor = FileDescriptor(path=file_path)
+                result = self.client.ingest_from_file(file_descriptor=file_descriptor, ingestion_properties=self.ingestion_props)
+                #if ingestion result is success or queued, then move the file to temp folder
+                if result.status == IngestionStatus.QUEUED or result.status == IngestionStatus.SUCCESS:
```

Comment:
there is some code commonality between the modules.Though they are not fully related, do you think it is a good idea to extract common code (just a thought)

Reply:
This module(splunk-adx-alert-addon) should not have been in the code commits of this branch. Deleted it.

Yes, I do agree there are commonalities between these modules. It will be definitely a good idea to have a common lib. Only concern will be to manually include the util/lib while building the splunk addon. Now all 3 of these modules(splunk-addon, splunk-forwarder-connector, splunk-spark-job) can independently built/deployed.

Reply:
how does multi module set  up in python. When creating the final upload package is it possible to include (like how we do in a multi module project in java/maven)

Reply:
that can be done in setup.py using sys.path.insert. I have added it as an issue of refactoring common code and will take care in the next release.